src/messages/proposals.rs

```diff
@@ -117,39 +134,47 @@
     }
 }
 
-impl Codec for ProposalID {
-    fn encode(&self, buffer: &mut Vec<u8>) -> Result<(), CodecError> {
-        encode_vec(VecSize::VecU8, buffer, &self.value)?;
-        Ok(())
-    }
-    fn decode(cursor: &mut Cursor) -> Result<Self, CodecError> {
-        let value = decode_vec(VecSize::VecU8, cursor)?;
-        Ok(ProposalID { value })
-    }
-}
-
 /// Alternative representation of a Proposal, where the sender is extracted from
 /// the encapsulating MLSPlaintext and the ProposalID is attached.
 #[derive(Clone)]
-pub struct QueuedProposal {
-    proposal: Proposal,
+pub struct QueuedProposal<'a> {
+    proposal: &'a Proposal,
     proposal_id: ProposalID,
     sender: Sender,
+    por_type: ProposalOrRefType,
 }
 
-impl QueuedProposal {
+impl<'a> QueuedProposal<'a> {
     /// Creates a new `QueuedProposal` from an `MLSPlaintext`
-    pub(crate) fn new(ciphersuite: &Ciphersuite, mls_plaintext: &MLSPlaintext) -> Self {
+    pub(crate) fn from_mls_plaintext(
+        ciphersuite: &Ciphersuite,
+        mls_plaintext: &'a MLSPlaintext,
+    ) -> Self {
         debug_assert!(mls_plaintext.content_type == ContentType::Proposal);
         let proposal = match &mls_plaintext.content {
-            MLSPlaintextContentType::Proposal(p) => p.clone(),
+            MLSPlaintextContentType::Proposal(p) => p,
             _ => panic!("API misuse. Only proposals can end up in the proposal queue"),
```

Comment:
Let's return an error instead of panicking. Now that the function isn't called `new` anymore it's ok to return a `Result`.

Reply:
Shouldn't it then be called `try_from`?